package.json

```diff
@@ -40,11 +40,11 @@
         "mkdirp": "^0.5.3",
         "resolve": "^1.3.2",
         "semver": "^5.3.0",
-        "tslib": "^1.10.0",
+        "tslib": "^2.0.0",
```

Comment:
Is there a reason to bump the major version of tslint? This would be a breaking change for some consumers, so ideally we should try to avoid it if possible.

Reply:
Why is this a breaking change? Tslib is bound to the TS version used to compile the library, ie: tslint. (Unless I am missing some other use-case where tslib is used in tslint)

Since TS 3.9, tslib version 2 is recommended, to avail of new features. More context about this here: https://github.com/microsoft/tslib/pull/109

That said, if no one reported any issue I can revert this, but technically it should be 2.0.0.

Reply:
Gotcha, thanks, that's helpful. The case I'm a little concerned about is a rule/plugin consumer accidentally relying on TSLint using a particular version of TSLint, though admittedly that is a pretty rare case.

@adidahiya, is this explicitly a concern or non-concern to your knowledge?

If no reasons to avoid the version are mentioned by anyone with historical context by the time TS 4.0 lands, I think we can safely merge.

Reply:
In that case the plugin author will be relying on perfect module hosting by package managers. Ie: the starts to align 😀

Reply:
I'm fairly confident that an upgrade to tslib 2.0.0 is __not__ required here. If we publish tslint built with tslib 2.0.0, then our consumers will be required to use TS 3.9+, and that means our `peerDependencies` typescript range is invalid. I just did a large upgrade to tslib 2.0.0 across many repos and it was fairly disruptive, so I would like to avoid that in this PR. TS 3.9+ should work fine with the latest tslib 1.x version.

Reply:
@adidahiya, can you explain why updating tslib here should effect consumers of tslint?

Tslib is a direct dependency of tslint and is bound to the TypeScript version used to **build** tslint.

> means our peerDependencies typescript range is invalid.
I don’t think this is correct, if the user uses tslib for their application they will need to explicitly have it as a devDependency and they choose the version based on their TS version.

> I just did a large upgrade to tslib 2.0.0 across many repos and it was fairly disruptive
Can you please elaborate? I’d love to know more. Updating to tslib 2 shouldn’t be distinctive at all, if you use TS 3.9+ to compile your project.

Reply:
Ps: I don’t have any problem removing this, just want to have more details.

Reply:
tslib 1.13 (_not_ 1.12) is compatible with TS <3.9 and 3.9+. So we should declare a dependency on that range, `~1.13.0`.

tslib is not simply used at build time, it is a runtime dependency. It provides helpers for TS language features which are not present in all JS runtime environments, for example object spread and `import *` syntax. So far, tslib and typescript have not declared an official "compatibility matrix" at runtime, and it was generally guaranteed that any version of tslib could work with any version of TS (unless otherwise stated). tslib 2.0 did a major version bump because they broke this contract, and any library which publishes a version which depends on tslib 2.0 at runtime _will not compile for consumers using TS <3.9_. Although it is true that most consumers of tslint simply use the CLI and don't care about the generated typings, there are TS consumers who use the programmatic API (e.g. custom rule plugins) and we should not break that API.

In my experience, when I did a large upgrade to tslib, I experienced some breaking changes in `import *` semantics at runtime which forced me to refactor a lot of code which modifying imports of external modules (specifically, in the Jest mocking system).

In any case, since tslint is deprecated, we should play it safe and not major bump our tslib dependency.

Reply:
Thanks @adidahiya for the detailed explanation. I’ll be doing the change tmr, as currently I am AFK.

Reply:
Tslib change done, though I still think this is incorrect. 

> tslib 2.0 did a major version bump because they broke this contract, and any library which publishes a version which depends on tslib 2.0 at runtime will not compile for consumers using TS <3.9. 

I don't think this is the case, tslib doesn't effect the emitted DTS hence it will not effect the compilation of the consumers of the library. But since tslint is deprecated I think it's okay to play it safe.